draft-ietf-jsonpath-base.md

```diff
@@ -1202,36 +1202,89 @@ The `filter-selector` works with arrays and objects exclusively. Its result is a
 A relative path, beginning with `@`, refers to the current array element or member value as the
 filter selector iterates over the array or object.
 
+##### Existence Tests
+{: unnumbered}
+
 A singular path by itself in a Boolean context is an existence test which yields true if the path selects a node and yields false if the path does not select a node.
-To be more specific about the actual value of a node selected by a path, explicit comparisons are necessary. This existence test — as an exception to the general rule — also works with nodes with structured values.
+This existence test — as an exception to the general rule — also works with nodes with structured values.
+
+To test the value of a node selected by a path, an explicit comparison is necessary.
+For example, to test whether the node selected by the path `@.foo` has the value `null`, use `@.foo == null` (see {{null-semantics}})
+rather than the negated existence test `!@.foo` (which yields false if `@.foo` selects a node, regardless of the node's value).
+
+##### Comparisons
+{: unnumbered}
 
 When a path resulting in an empty nodelist appears on either side of a comparison, the comparison yields
-true if and only if the comparison operator is `==`, `>=` or `<=` and the other side of the comparison is also a path
-resulting in an empty nodelist.
+true if and only if:
 
-When no path resulting in an empty nodelist appears on either side of a comparison, any path which appears on either
-side of the comparison and results in a nodelist consisting of a single node is replaced by the value of the node
-and then:
+* the comparison operator is `==`, `>=` or `<=` and the other side of the comparison is also a path
+resulting in an empty nodelist, or
+* the comparison operator is `!=` and the other side of the comparison is not also a path resulting in an empty nodelist.
 
-* comparison using one of the operators `==` or `!=` yields true if and only if the comparison
-is between:
-    * primitive values which satisfy the comparison, or
-    * structured values compared using `!=`.
+When any path on either side of a comparison results in a nodelist consisting of a single node, each such path is
+replaced by the value of its node and then:
 
-* comparisons using one of the operators `<`, `<=`, `>`, or `>=` yield true if and only if
+* comparison using one of the operators `==` yields true if and only if the comparison
+is between equal primitive values.
+
+* comparisons using one of the operators `<=` or `>=` yield true if and only if
 the comparison is between numeric values which satisfy the comparison.
 
-Note that comparisons between structured values, even if the values are equal, yield false.
+* any comparison of two values using one of the operators `!=`, `>`, `<` is defined as the negation of the comparison
+of the same values using the operator `==`, `<=`, `>=`, respectively.
+
+* data types are not implicitly converted in comparisons.
```

Comment:
Oops, I meant the last bullet point "data types are not implicitly converted in comparisons" - but my suggestion stands: there's no real need to bring the notion of "conversion" in, just assert that all type-incompatible comparisons, and all comparisons involving structured values, return false.

Reply:
> This is getting nicer. But I'm not crazy about this title, because what people care about isn't "conversion", it's the result. How about text something along the lines of "All comparisons between values which are not of the same JSON type, or where either or both values are structured, return false."

There is a simple solution: delete the offending bullet. The other bullets imply that comparison of values with distinct JSON types yield false. I've preserved an example of this and changed the comment to talk about type mismatch. Please see ef71e3ef709496e12d49237f7e6e1c3017dee9bf.

> 
> Perhaps the phrase "JSON type" should be footnoted with a reference to https://datatracker.ietf.org/doc/html/rfc8259#section-3

That section doesn't say much about JSON types, but the introduction does:
> JSON can represent four primitive types (strings, numbers, booleans,
   and null) and two structured types (objects and arrays).

Our spec uses the term "type" (meaning JSON type) in many places. For the avoidance of doubt, I added the term in b0c9f2dd2e79da499606e6f62a403f554818a8e4.

Reply:
> "All comparisons between values which are not of the same JSON type, or where either or both values are structured, return false."

That would not work very well, as you no longer could compare with null (or any other value used as a stand-in value).